### DISCUSSION TITLE
Place tooltip bottom-start from pointer not element?

### DISCUSSION THREAD

**Discussion by viveleroi** [ReactTooltip/react-tooltip#1159] [Q&A]
Is there a way to place the tooltip bottom-start from the *mouse pointer* location? It looks like the tooltip attaches to the element and doesn't move, but our design requirements expect the tooltip to float below and to the right of the pointer.

![chrome_47hEYyFWfp](https://github.com/ReactTooltip/react-tooltip/assets/70734/16a55e5c-f333-4c2f-97d1-c89c5299bbcd)

I thought the imperative ref "api" might work but I don't see a way to set custom position logic "on open".

**Comment by gabrieljablonski** [comment]
I don't believe placing something relative to the bottom of the pointer is a straight-forward design for browsers. Default cursor sizes may vary between different operating systems (though 32x32 seems like the norm), and the user may have a custom cursor as well, so the experience might be different across users. Also keeping track of the page zoom to adjust the offset is definitely doable, but a bit annoying.

With that out of the way, assuming you still want to try keeping this design, while the tooltip does offer the `float` prop, which makes the tooltip follow the pointer, and you can you combine it with the `offset` prop to achieve that effect (though I use the `middlewares` prop instead in the example, explanation below).

Here's the example:

https://codesandbox.io/p/sandbox/react-tooltip-overflow-forked-ptgn7n?file=%2Fsrc%2FApp.js%3A34%2C17

It will look broken on this page, since the application runs inside an iframe, but it should look correct if you open the "standalone" demo application here:

https://ptgn7n.csb.app/

On a further note, the tooltip by default flips from bottom to top (or top to bottom) when it would overflow the viewport. Here's how it looks:

https://github.com/ReactTooltip/react-tooltip/assets/21102974/e5a9b5f7-6fbe-472b-8a60-b4a023e069d7

So that doesn't happen, we overwrite the default middlewares and omit `flip()`.

---

At the end of the day, from a design/UX perspective I'd still recommend either keeping the tooltip with the default `float` positioning (above pointer), or have a regular tooltip anchored to the input.

**Reply by viveleroi** [reply]
Thanks for your thorough answer, I really appreciate it. Our application is for a specific organization and we don't quite have to worry about cross-OS/browser support, zooms, and custom pointers. I personally prefer the tooltips being above the element but I think our design teams are trying to replicate native tooltip behavior.

I will give this a try and get back to you

**Reply by gabrieljablonski** [reply]
> I personally prefer the tooltips being above the element but I think our design teams are trying to replicate native tooltip behavior.

I know the feeling haha 😅

Let me know if this works then.

**Reply by viveleroi** [reply]
Looks like it does position like I want and it follows the cursor. Is it possible to retain that initial position, but not follow the cursor? I realized after further reading our design docs and looking at older apps, the tooltip appears offset from the cursor like this, but it doesn't follow the cursor. It stays at this position until the mouse leaves the element or the timeout of 10 seconds expires and it closes.

**Reply by gabrieljablonski** [reply]
> Is it possible to retain that initial position, but not follow the cursor?

`float` and `offset` won't help you then. You'll need to control the position manually. See our `"onClickAnchor"` example on the [dev test demo](https://github.com/ReactTooltip/react-tooltip/blob/master/src/App.tsx#L138-L159) (can be run by cloning the repo and using `yarn dev`), or the docs for how to use [the imperative mode](https://react-tooltip.com/docs/examples/imperative-mode).

Keep in mind the dev test demo uses some deprecated props, but the deprecation message should point you to the correct props.

**Reply by viveleroi** [reply]
That example attaches the `handlePositionClick` event handler to the button itself. We're going to have tooltips in dozens of components so I'll need to think of the best way to handle this. 

I could probably add a mouseenter event listener on the body and check if the target is a tooltip element, or I could create a custom hook that exposes a function I could assign to the onMouseEnter prop. Not sure which is best - the global listener is less code but technically would fire _a lot_, though luckily anything not matching `[data-tooltip-id]` would be ignored

**Reply by viveleroi** [reply]
Oh right, a global listener won't work because `mouseenter` is not an event that bubbles. I have no choice but to add a custom listener to every element we want tooltips for.

**Reply by viveleroi** [reply]
Ok I decide to try a custom middleware and it seems to work fine. I track `mousemove` and store the position. When the middleware is called I use that position and cache it, so that I can continue to use that first position until I clear the cached value on `afterHide`.

This actually seemed to work decently well but has two problems: we set `delayShow=1000` which causes the position to be wrong by the time it opens, and the positions are wrong when inside of an iframe which breaks our storybook and we'll definitely use the app inside of iframes sadly at some future time.

Not sure what to do from here, lol.

**Reply by gabrieljablonski** [reply]
At this point you might wanna look into some other tooltip solutions. There are some CSS-only alternatives that might work for your use-case out of the box.

It is definitely possible to tweak `react-tooltip` to work exactly as you want, but it might not be worth the trouble.

**Reply by viveleroi** [reply]
I'm close enough I will probably try to continue. CSS-only probably won't work well enough given the complexity of our requirements and so far I haven't found a library that clearly handles this use case while also supporting enough for our other requirements.

If you have any ideas on why middleware approaches inside of an iframe break, please let me know so I can get a jump start. Not sure what to do about a delay in showing either but that seems easier to fix.

**Reply by viveleroi** [reply]
Well my approach breaks storybook because it seems like they're using `transform/scale` on elements which reset a position: fixed element's container from the viewport. Not sure I can fix that but that's less of a concern, as it works fine outside of the auto-generated docs.

**Reply by viveleroi** [reply]
Is it possible to render the tooltip in a portal? We already have portals we could use for this and that would very likely solve our positioning problem

**Reply by viveleroi** [reply]
I think positioning the tooltip via the mouse is not great UX especially when we're dealing with form fields, it'll cover the element and might block the user. I'm going to go back to our requirements people and push back on that.
